BLD: Fails to build: internal_error

### Bug report

**Bug summary**

When building the package as part of the packaging `yt` for openSUSE, the fails often (not always, something between 50 % and 80 %) it fails with an error like this:

```
[  338s]   gcc -Wno-unused-result -Wsign-compare -DNDEBUG -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=r
eturn-type -g -DOPENSSL_LOAD_CONF -fwrapv -fno-semantic-interposition -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protecti
on -Werror=return-type -g -IVendor/ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -g -IVendor
/ -O2 -Wall -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=3 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -freport-bug -fPIC -Iyt/utilities/
lib/ -Iyt/utilities/lib -I./yt/utilities/lib -I/usr/include/python3.10 -I/usr/lib64/python3.10/site-packages/numpy/core/include -c yt/utilities/lib/marching_cubes.cpp -o build/temp.linux-x86_64-cpython-310/yt/u
tilities/lib/marching_cubes.o
[  338s]   lto1: internal compiler error: resolution sub id 0xadc1906a211a636c not in object file
[  338s]   0x102f76b internal_error(char const*, ...)
[  338s]         ???:0
[  338s]   0x15cfb50 lto_main()
[  338s]         ???:0
[  338s]   Please submit a full bug report, with preprocessed source.
[  338s]   Please include the complete backtrace with any bug report.
[  338s]   See <https://bugs.opensuse.org/> for instructions.
[  338s]   lto-wrapper: fatal error: /usr/bin/g++ returned 1 exit status
[  338s]   compilation terminated.
[  338s]   /usr/lib64/gcc/x86_64-suse-linux/12/../../../../x86_64-suse-linux/bin/ld: error: lto-wrapper failed
[  338s]   collect2: error: ld returned 1 exit status
```

**Code for reproduction**

It gets triggered while building the package, so no obvious code snippet.

[Complete build log](https://github.com/yt-project/yt/files/10331308/_log.txt) with all packages used and steps taken to achieve this problem.

**Actual outcome**

See the build error above.

**Expected outcome**

Clean build.

**Version Information**
  * Operating System: Linux/openSUSE/Tumbleweed (rolling distro) as of 2023-01-02
  * Python Version: this one is for Python 3.10.9, but I don’t think it is interpreter driven
  * yt version: 4.1.2 from the PyPI tarball
  * Other Libraries (if applicable): see above log for the complete list

All packages from openSUSE packages.

----

When I have consulted this with our GCC expert (hi, @marxin!) I got this suggestion:

> Hi, I'm almost sure that the problem is that there are some wrong dependencies and some .o file is used before its producer completes it. I have seen it many times, it manifests itself in different voices and the fact that it is not deterministic:
> 
> [   10s]   lto1: internal compiler error: resolution sub id 0xb2bc78268f7df7a0 not in object file
> [   10s]   0x102f987 internal_error(char const*, ...)
> [   10s]         ???:0
> [   10s]   0x15d02f0 lto_main()
> [   10s]         ???:0
> 
> or that message of yours with lto1: internal compiler error: original not compressed with zstd

--- a/setup.py
+++ b/setup.py
@@ -11,6 +11,7 @@ from setupext import (
     check_for_openmp,
     check_for_pyembree,
     create_build_ext,
+    get_python_include_dirs,
     install_ccompiler,
 )
 
@@ -40,6 +41,8 @@ else:
 CPP14_FLAG = CPP14_CONFIG[_COMPILER]
 CPP11_FLAG = CPP11_CONFIG[_COMPILER]
 
+FIXED_INTERP = "fixed_interpolator"
+
 cythonize_aliases = {
     "LIB_DIR": "yt/utilities/lib/",
     "LIB_DIR_GEOM": ["yt/utilities/lib/", "yt/geometry/"],
@@ -52,7 +55,7 @@ cythonize_aliases = {
     "EWAH_LIBS": std_libs
     + [os.path.abspath(importlib_resources.files("ewah_bool_utils"))],
     "OMP_ARGS": omp_args,
-    "FIXED_INTERP": "yt/utilities/lib/fixed_interpolator.cpp",
+    "FIXED_INTERP": FIXED_INTERP,
     "ARTIO_SOURCE": sorted(glob.glob("yt/frontends/artio/artio_headers/*.c")),
     "CPP14_FLAG": CPP14_FLAG,
     "CPP11_FLAG": CPP11_FLAG,
@@ -88,8 +91,32 @@ class BinaryDistribution(Distribution):
 
 
 if __name__ == "__main__":
+    # Avoid a race condition on fixed_interpolator.o during parallel builds by
+    # building it only once and storing it in a static library.
+    # See https://github.com/yt-project/yt/issues/4278 and
+    # https://github.com/pypa/setuptools/issues/3119#issuecomment-2076922303
+    # for the inspiration for this fix.
+
+    # build_clib doesn't add the Python include dirs (for Python.h) by default,
+    # as opposed to build_ext, so we need to add them manually.
+    clib_include_dirs = get_python_include_dirs()
+
+    # fixed_interpolator.cpp uses Numpy types
+    import numpy
+
+    clib_include_dirs.append(numpy.get_include())
+
+    fixed_interp_lib = (
+        FIXED_INTERP,
+        {
+            "sources": ["yt/utilities/lib/fixed_interpolator.cpp"],
+            "include_dirs": clib_include_dirs,
+        },
+    )
+
     setup(
         cmdclass={"sdist": sdist, "build_ext": build_ext},
         distclass=BinaryDistribution,
+        libraries=[fixed_interp_lib],
         ext_modules=[],  # !!! We override this inside build_ext above
     )

--- a/setupext.py
+++ b/setupext.py
@@ -9,6 +9,7 @@ import sys
 import tempfile
 from textwrap import dedent
 from concurrent.futures import ThreadPoolExecutor
+from distutils import sysconfig
 from distutils.ccompiler import CCompiler, new_compiler
 from distutils.sysconfig import customize_compiler
 from subprocess import PIPE, Popen
@@ -366,6 +367,31 @@ def install_ccompiler():
         return objects
 
     CCompiler.compile = _compile
+
+
+def get_python_include_dirs():
+    """Extracted from distutils.command.build_ext.build_ext#finalize_options(),
+    https://github.com/python/cpython/blob/812245ecce2d8344c3748228047bab456816180a/Lib/distutils/command/build_ext.py#L148-L167
+    """
+    include_dirs = []
+
+    # Make sure Python's include directories (for Python.h, pyconfig.h,
+    # etc.) are in the include search path.
+    py_include = sysconfig.get_python_inc()
+    plat_py_include = sysconfig.get_python_inc(plat_specific=1)
+
+    # If in a virtualenv, add its include directory
+    # Issue 16116
+    if sys.exec_prefix != sys.base_exec_prefix:
+        include_dirs.append(os.path.join(sys.exec_prefix, 'include'))
+
+    # Put the Python "system" include dir at the end, so that
+    # any local include dirs take precedence.
+    include_dirs.extend(py_include.split(os.path.pathsep))
+    if plat_py_include != py_include:
+        include_dirs.extend(plat_py_include.split(os.path.pathsep))
+
+    return include_dirs
 
 
 def create_build_ext(lib_exts, cythonize_aliases):

--- a/setupext.py
+++ b/setupext.py
@@ -370,7 +370,7 @@ def install_ccompiler():
 
 
 def get_python_include_dirs():
-    """Extracted from distutils.command.build_ext.build_ext#finalize_options(),
+    """Extracted from distutils.command.build_ext.build_ext.finalize_options(),
     https://github.com/python/cpython/blob/812245ecce2d8344c3748228047bab456816180a/Lib/distutils/command/build_ext.py#L148-L167
     """
     include_dirs = []

--- a/yt/utilities/lib/grid_traversal.pyx
+++ b/yt/utilities/lib/grid_traversal.pyx
@@ -1,5 +1,5 @@
 # distutils: include_dirs = LIB_DIR
-# distutils: libraries = STD_LIBS FIXED_INTERP
+# distutils: libraries = STD_LIBS
 # distutils: language = c++
 # distutils: extra_compile_args = CPP14_FLAG
 # distutils: extra_link_args = CPP14_FLAG
@@ -18,8 +18,6 @@ cimport numpy as np
 from libc.math cimport atan2, cos, fabs, floor, sin, sqrt
 
 from yt.utilities.lib.fp_utils cimport fmin
-
-from .fixed_interpolator cimport *
 
 
 @cython.boundscheck(False)

--- a/yt/utilities/lib/grid_traversal.pyx
+++ b/yt/utilities/lib/grid_traversal.pyx
@@ -1,6 +1,5 @@
 # distutils: include_dirs = LIB_DIR
-# distutils: libraries = STD_LIBS
-# distutils: sources = FIXED_INTERP
+# distutils: libraries = STD_LIBS FIXED_INTERP
 # distutils: language = c++
 # distutils: extra_compile_args = CPP14_FLAG
 # distutils: extra_link_args = CPP14_FLAG

--- a/yt/utilities/lib/image_samplers.pyx
+++ b/yt/utilities/lib/image_samplers.pyx
@@ -1,8 +1,7 @@
 # distutils: include_dirs = LIB_DIR
 # distutils: extra_compile_args = CPP14_FLAG OMP_ARGS
 # distutils: extra_link_args = CPP14_FLAG OMP_ARGS
-# distutils: libraries = STD_LIBS
-# distutils: sources = FIXED_INTERP
+# distutils: libraries = STD_LIBS FIXED_INTERP
 # distutils: language = c++
 """
 Image sampler definitions

--- a/yt/utilities/lib/marching_cubes.pyx
+++ b/yt/utilities/lib/marching_cubes.pyx
@@ -1,6 +1,5 @@
 # distutils: include_dirs = LIB_DIR
-# distutils: libraries = STD_LIBS
-# distutils: sources = FIXED_INTERP
+# distutils: libraries = STD_LIBS FIXED_INTERP
 # distutils: language = c++
 """
 Marching cubes implementation

--- a/yt/utilities/lib/partitioned_grid.pyx
+++ b/yt/utilities/lib/partitioned_grid.pyx
@@ -1,6 +1,5 @@
-# distutils: sources = FIXED_INTERP
 # distutils: include_dirs = LIB_DIR
-# distutils: libraries = STD_LIBS
+# distutils: libraries = STD_LIBS FIXED_INTERP
 # distutils: language = c++
 # distutils: extra_compile_args = CPP14_FLAG
 # distutils: extra_link_args = CPP14_FLAG

